pkg/ccl/backupccl/alter_backup_planning.go

```diff
@@ -0,0 +1,265 @@
+		return errors.Wrapf(err, "failed to open backup storage location")
+	}
+	defer baseStores.Close()
+
+	opts, err := readEncryptionOptions(ctx, baseStores)
+	if err != nil {
+		return err
+	}
+
+	ioConf := baseStores.ExternalIOConf()
+
+	// Check that at least one of the old keys works on the backup.
+	// Use the first one that works to decrypt the ENCRYPTION-INFO file(s).
+	var defaultKMSInfo *jobspb.BackupEncryptionOptions_KMSInfo
+	oldKMSFound := false
+	for i := len(oldKms) - 1; i >= 0; i-- {
+		for _, encFile := range opts {
+			defaultKMSInfo, err = validateKMSURIsAgainstFullBackup([]string{oldKms[i]},
+				newEncryptedDataKeyMapFromProtoMap(encFile.EncryptedDataKeyByKMSMasterKeyID), &backupKMSEnv{
+					baseStores.Settings(),
+					&ioConf,
+				})
+
+			if err == nil {
+				oldKMSFound = true
+				break
+			}
+		}
+		if oldKMSFound {
+			break
+		}
+	}
+	if !oldKMSFound {
+		return fmt.Errorf("no key in old_kms matches the provided backup")
+	}
+
+	encryption := &jobspb.BackupEncryptionOptions{
+		Mode:    jobspb.EncryptionMode_KMS,
+		KMSInfo: defaultKMSInfo}
+
+	// Recover the encryption key using the old key, so we can encrypt it again with the new keys.
+	var plaintextDataKey []byte
+	plaintextDataKey, err = getEncryptionKey(ctx, encryption, baseStores.Settings(),
+		baseStores.ExternalIOConf())
+
+	if err != nil {
+		return err
+	}
+
+	kmsEnv := &backupKMSEnv{settings: p.ExecCfg().Settings, conf: &p.ExecCfg().ExternalIODirConfig}
+
+	encryptedDataKeyByKMSMasterKeyID := newEncryptedDataKeyMap()
+
+	// Add each new key user wants to add to a new data key map.
+	for _, kmsURI := range newKms {
+		masterKeyID, encryptedDataKey, err := getEncryptedDataKeyFromURI(ctx,
+			plaintextDataKey, kmsURI, kmsEnv)
+		if err != nil {
+			return err
+		}
+
+		encryptedDataKeyByKMSMasterKeyID.addEncryptedDataKey(plaintextMasterKeyID(masterKeyID),
+			encryptedDataKey)
+	}
+
+	encryptedDataKeyMapForProto := make(map[string][]byte)
+	encryptedDataKeyByKMSMasterKeyID.rangeOverMap(
+		func(masterKeyID hashedMasterKeyID, dataKey []byte) {
+			encryptedDataKeyMapForProto[string(masterKeyID)] = dataKey
+		})
+
+	encryptionInfo := &jobspb.EncryptionInfo{EncryptedDataKeyByKMSMasterKeyID: encryptedDataKeyMapForProto}
+
+	// Write the new ENCRYPTION-INFO file.
+	err = writeAlteredEncryptionInfo(ctx, encryptionInfo, baseStores)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func getEncryptionInfoFiles(ctx context.Context, dest cloud.ExternalStorage) ([]string, error) {
+	var files []string
+	// Look for all files in dest that start with "/ENCRYPTION-INFO" and return them.
+	err := dest.List(ctx, "", "", func(p string) error {
+		paths := strings.Split(p, "/")
+		p = paths[len(paths)-1]
+		if match := strings.HasPrefix(p, backupEncryptionInfoFile); match {
+			files = append(files, p)
+		}
+
+		return nil
+	})
+	if len(files) < 1 {
+		return nil, errors.Errorf("no encrypted files found")
+	}
+
+	return files, err
+}
+
+func writeAlteredEncryptionInfo(
```

Comment:
nit: rename to `writeNewEncryptionInfoToBackup`

can we just pass in the new file name from the doPlanHook method? That way we don't have to call `getEncryptionInfoFiles` again, cause we already do that when checking the oldKMSURIs.